### PR TITLE
Support fx_pytree in dynamo

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105574

This PR does two things:
1. Make dynamo trace through fx_pytree (on top of torch.utils._pytree) so that generated graph modules can be retraced. 
2. Fix bug where unflatten not returning dynamo VariableTracker. 

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

Differential Revision: [D47734623](https://our.internmc.facebook.com/intern/diff/D47734623)

### CODE REVIEW

File: torch/_dynamo/variables/user_defined.py [hunk 1]
```diff
@@ -244,6 +249,17 @@ def call_method(
...
+                method is dict.__contains__
+                and len(args) == 1
+                and isinstance(args[0], (ConstantVariable, BuiltinVariable))
+                and inspect.getattr_static(type(self.value), "keys") is dict.keys
+            ):
+                assert not kwargs
+                return ConstantVariable(
+                    args[0].as_python_constant() in self.value, **options
+                ).add_guard(self.source.make_guard(GuardBuilder.ODICT_KEYS))
+
```
Comment (yanboliang) [inline torch/_dynamo/variables/user_defined.py:262]: L240-L262 are pretty similar logic for ```dict``` and ```OrderedDict```, can you consolidate them?

**Review by yanboliang — APPROVED** [review]

File: torch/_dynamo/variables/user_defined.py [hunk 2]
```diff
@@ -235,7 +240,7 @@ def call_method(
             if (
                 method is collections.OrderedDict.__contains__
                 and len(args) == 1
-                and isinstance(args[0], ConstantVariable)
+                and isinstance(args[0], (ConstantVariable, BuiltinVariable))
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply torch/_dynamo/variables/user_defined.py:243]: Yep

[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply torch/_dynamo/variables/user_defined.py:243]: To make sure something like this works:
```py
look_up_dict = {tuple: 4}
a = (5, 6)
type(a) in look_up_dict
```

File: test/dynamo/test_export.py [hunk 3]
```diff
@@ -3237,6 +3237,89 @@ def forward(self, l_x_, ones_3_true_branch, ones_1_true_branch, ones_true_branch
     return add_2""",
         )
 
+    def test_fx_pytree(self):
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply test/dynamo/test_export.py:3240]: Sure

File: torch/_dynamo/variables/user_defined.py (hunk 1 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/variables/user_defined.py:262]: Yep!

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47734623).

File: test/dynamo/test_export.py [hunk 4]
```diff
@@ -3303,6 +3303,102 @@ def forward(self, l_x_, ones_3_true_branch, ones_1_true_branch, ones_true_branch
...
+        self.assertTrue(torch.allclose(gm(inp_test)["b"], gm2(inp_test)["b"]))
+
+    def test_fx_pytree(self):
+        def foo(args):
+            flat_args, spec = torch.utils._pytree.tree_flatten(args)
+            flat_args_fx = torch.fx._pytree.tree_flatten_spec(args, spec)
+            return flat_args_fx[0] + flat_args[0]
+
+        inp_container = (torch.randn(20, 16, 50, 100), torch.randn(20, 16, 50, 100))
+        print(foo(inp_container))
```
Comment (ydwu4) [inline test/dynamo/test_export.py:3396]: nit: remove the print

**Review by ydwu4 — APPROVED** [review]

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47734623).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47734623).

**Pull request closed without merging**